independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java

```diff
@@ -839,16 +833,29 @@ protected void implementDestroy(BeanInfo bean, ClassCreator beanCreator, Provide
                                     callback.declaringClass().name(), callback.name()));
                         }
                         reflectionRegistration.registerMethod(callback);
-                        destroy.invokeStaticMethod(MethodDescriptors.REFLECTIONS_INVOKE_METHOD,
-                                destroy.loadClass(callback.declaringClass().name().toString()),
-                                destroy.load(callback.name()), destroy.newArray(Class.class, destroy.load(0)),
+                        preDestroyBytecode.invokeStaticMethod(MethodDescriptors.REFLECTIONS_INVOKE_METHOD,
+                                preDestroyBytecode.loadClass(callback.declaringClass().name().toString()),
+                                preDestroyBytecode.load(callback.name()),
+                                preDestroyBytecode.newArray(Class.class, preDestroyBytecode.load(0)),
                                 destroy.getMethodParam(0),
-                                destroy.newArray(Object.class, destroy.load(0)));
+                                preDestroyBytecode.newArray(Object.class, preDestroyBytecode.load(0)));
                     } else {
                         // instance.superCoolDestroyCallback()
-                        destroy.invokeVirtualMethod(MethodDescriptor.of(callback), destroy.getMethodParam(0));
+                        preDestroyBytecode.invokeVirtualMethod(MethodDescriptor.of(callback), destroy.getMethodParam(0));
                     }
                 }
+                preDestroyBytecode.returnVoid();
+
+                // PreDestroy interceptors
+                if (!bean.getLifecycleInterceptors(InterceptionType.PRE_DESTROY).isEmpty()) {
+                    destroy.invokeVirtualMethod(
+                            MethodDescriptor.ofMethod(SubclassGenerator.generatedName(bean.getProviderType().name(), baseName),
+                                    SubclassGenerator.DESTROY_METHOD_NAME, void.class, Runnable.class),
+                            destroy.getMethodParam(0), preDestroyForwarder.getInstance());
+                } else {
+                    destroy.invokeInterfaceMethod(MethodDescriptor.ofMethod(Runnable.class, "run", void.class),
+                            preDestroyForwarder.getInstance());
```

Comment:
I wonder if it would make sense to optimize for the use case where only the target class defines a callback (which could be very common) and so the `Runnable` is just useless...

Reply:
I originally attempted that, because I thought of the exact same thing, but I failed. I'll try again.

Reply:
(Obviously I could just duplicate the code, that would work. I tried to be smarter, and _that_ failed :-) )